File: JavaScript/7-Trees/Algorithms/4-LowestCommonAncestor(LCA)/LowestCommonAncestor(LCA).js
// Definition of a binary tree node (used across all approaches)
function TreeNode(val) {
    this.val = val;
    this.left = null;
    this.right = null;
}

// Helper function to build a sample binary tree for testing
function buildSampleTree() {
    const root = new TreeNode(3);
    root.left = new TreeNode(5);
    root.right = new TreeNode(1);
    root.left.left = new TreeNode(6);
    root.left.right = new TreeNode(2);
    root.right.left = new TreeNode(0);
    root.right.right = new TreeNode(8);
    root.left.right.left = new TreeNode(7);
    root.left.right.right = new TreeNode(4);
    return root;
}

let treeRoot = buildSampleTree(); // Build the tree once, use it for all methods
let p = treeRoot.left; // Node with value 5
let q = treeRoot.right; // Node with value 1
let p2 = treeRoot.left.left; // Node with value 6
let q2 = treeRoot.left.right.right; // Node with value 4

// 1. Recursive Approach
//   - Find paths to p and q, then compare them to find the last common node.
function findPath(root, target, path) {
    if (!root) return false;

    path.push(root);

    if (root === target) return true;

    if (findPath(root.left, target, path) || findPath(root.right, target, path)) {
        return true;
    }

    path.pop();
    return false;
}

function lowestCommonAncestorRecursive(root, p, q) {
    const pathP = [];
    const pathQ = [];

    findPath(root, p, pathP);
    findPath(root, q, pathQ);

    let lca = null;
    let i = 0;
    while (i < pathP.length && i < pathQ.length && pathP[i] === pathQ[i]) {
        lca = pathP[i];
        i++;
    }
    return lca;
}

console.log("1. Recursive Approach:");
console.log("LCA (5, 1):", lowestCommonAncestorRecursive(treeRoot, p, q).val); // Output: 3
console.log("LCA (6, 4):", lowestCommonAncestorRecursive(treeRoot, p2, q2).val); // Output: 5

// 2. Iterative Approach using Parent Pointers
//   - Use a stack to traverse the tree, keeping track of parent nodes.
//   - Store parent pointers for each node in a map.
//   - Find the paths to p and q using the parent pointers.
//   - Find the LCA by comparing the paths.

function lowestCommonAncestorIterative(root, p, q) {
    const stack = [root];
    const parentMap = new Map();
    parentMap.set(root, null); // Parent of root is null

    // Iterate until we find both p and q
    while (!parentMap.has(p) || !parentMap.has(q)) {
        const node = stack.pop();

        if (node.left) {
            parentMap.set(node.left, node);
            stack.push(node.left);
        }
        if (node.right) {
            parentMap.set(node.right, node);
            stack.push(node.right);
        }
    }

    // Get the path to p
    const pathP = [];
    while (p) {
        pathP.push(p);
        p = parentMap.get(p);
    }

    // Get the path to q
    const pathQ = [];
    while (q) {
        pathQ.push(q);
        q = parentMap.get(q);
    }

    // Find the LCA
    let lca = null;
    let i = pathP.length - 1;
    let j = pathQ.length - 1;
    while (i >= 0 && j >= 0 && pathP[i] === pathQ[j]) {
        lca = pathP[i];
        i--;
        j--;
    }
    return lca;
}

console.log("\n2. Iterative Approach using Parent Pointers:");
console.log("LCA (5, 1):", lowestCommonAncestorIterative(treeRoot, p, q).val); // Output: 3
console.log("LCA (6, 4):", lowestCommonAncestorIterative(treeRoot, p2, q2).val); // Output: 5

// 3. Optimized Recursive Approach (Without Path Storage)
//    - A more efficient recursive approach that doesn't require storing paths.
//    - If the current node is either p or q, or if the LCA is found in one subtree,
//      return the current node.
function lowestCommonAncestorOptimized(root, p, q) {
    if (!root || root === p || root === q) {
        return root;
    }

    const leftLCA = lowestCommonAncestorOptimized(root.left, p, q);
    const rightLCA = lowestCommonAncestorOptimized(root.right, p, q);

    if (leftLCA && rightLCA) {
        return root; // Found LCA: p and q are in different subtrees
    }

    return leftLCA || rightLCA; // p and q are in the same subtree, or one is the ancestor of the other
}

console.log("\n3. Optimized Recursive Approach:");
console.log("LCA (5, 1):", lowestCommonAncestorOptimized(treeRoot, p, q).val); // Output: 3
console.log("LCA (6, 4):", lowestCommonAncestorOptimized(treeRoot, p2, q2).val); // Output: 5

// 4. Iterative Approach using Morris Traversal
//    -  A tree traversal without using stack or recursion.  It's more complex.
//    -  Find LCA of two nodes in a binary tree using Morris traversal.
function lowestCommonAncestorMorris(root, p, q) {
  let current = root;
  let pAncestor = null;
  let qAncestor = null;

  // Function to find a node and store its ancestors
  const findNodeAndAncestors = (target) => {
    current = root;
    let ancestor = null;
    while (current) {
      if (current === target) {
        ancestor = current;
        break;
      } else if (target.val < current.val) {
        ancestor = current;
        current = current.left;
      } else {
        ancestor = current;
        current = current.right;
      }
    }
    return ancestor;
  };

  pAncestor = findNodeAndAncestors(p);
  qAncestor = findNodeAndAncestors(q);

  if (!pAncestor || !qAncestor) return null; // Handle cases where p or q is not in the tree

  const getPath = (node) => {
        const path = [];
        current = root;
        while(current && current != node){
            path.push(current);
            if(node.val < current.val){
                current = current.left;
            }
            else{
                current = current.right;
            }
        }
        path.push(node);
        return path;
  }
  const pathP = getPath(p);
  const pathQ = getPath(q);

  let lca = null;
  let i = 0;
  while (i < pathP.length && i < pathQ.length && pathP[i] === pathQ[i]) {
    lca = pathP[i];
    i++;
  }
  return lca;
}

console.log("\n4. Iterative Approach using Morris Traversal:");
console.log("LCA (5, 1):", lowestCommonAncestorMorris(treeRoot, p, q).val);
console.log("LCA (6, 4):", lowestCommonAncestorMorris(treeRoot, p2, q2).val);

// 5. Using Lowest Common Ancestor of a Binary Tree III (with parent pointers)
//    -  This approach is specifically for trees where each node has a parent pointer.
//    -  The provided example tree does not have parent pointers, so we'll create a new one.
//    -  This is NOT a standard LCA solution, but a solution to a specific problem.
function TreeNodeWithParent(val) {
        this.val = val;
        this.left = null;
        this.right = null;
        this.parent = null;
}

function buildTreeWithParentPointers() {
    const root = new TreeNodeWithParent(3);
    const node5 = new TreeNodeWithParent(5);
    const node1 = new TreeNodeWithParent(1);
    const node6 = new TreeNodeWithParent(6);
    const node2 = new TreeNodeWithParent(2);
    const node0 = new TreeNodeWithParent(0);
    const node8 = new TreeNodeWithParent(8);
    const node7 = new TreeNodeWithParent(7);
    const node4 = new TreeNodeWithParent(4);

    root.left = node5;
    root.right = node1;
    node5.left = node6;
    node5.right = node2;
    node1.left = node0;
    node1.right = node8;
    node2.left = node7;
    node2.right = node4;

    node5.parent = root;
    node1.parent = root;
    node6.parent = node5;
    node2.parent = node5;
    node0.parent = node1;
    node8.parent = node1;
    node7.parent = node2;
    node4.parent = node2;

    return root;
}
function lowestCommonAncestorWithParentPointers(p, q) {
    const pathP = [];
    const pathQ = [];

    // Traverse up to the root from p and q, storing the path
    while (p) {
        pathP.push(p);
        p = p.parent;
    }
    while (q) {
        pathQ.push(q);
        q = q.parent;
    }

    // Find the common ancestor
    let lca = null;
    let i = pathP.length - 1;
    let j = pathQ.length - 1;
    while (i >= 0 && j >= 0 && pathP[i] === pathQ[j]) {
        lca = pathP[i];
        i--;
        j--;
    }
    return lca;
}
let treeRootWithParent = buildTreeWithParentPointers();
let pWithParent = treeRootWithParent.left;  // Node 5
let qWithParent = treeRootWithParent.right; // Node 1
let p2WithParent = treeRootWithParent.left.left; // Node 6
let q2WithParent = treeRootWithParent.left.right.right;  // Node 4

console.log("\n5. Using Parent Pointers:");
console.log("LCA (5, 1):", lowestCommonAncestorWithParentPointers(pWithParent, qWithParent).val);
console.log("LCA (6, 4):", lowestCommonAncestorWithParentPointers(p2WithParent, q2WithParent).val);

File: C++/7-Trees/Algorithms/4-LowestCommonAncestor(LCA)/LowestCommonAncestor(LCA).cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <algorithm>
#include <map>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    // Approach 1: Recursive Approach
    // - Recursively search for p and q in the tree.
    // - If both p and q are found in the left subtree, the LCA must be in the left subtree.
    // - If both p and q are found in the right subtree, the LCA must be in the right subtree.
    // - Otherwise, the current node is the LCA.
    TreeNode* lowestCommonAncestorRecursive(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (root == NULL || root == p || root == q) {
            return root;
        }
        TreeNode* left = lowestCommonAncestorRecursive(root->left, p, q);
        TreeNode* right = lowestCommonAncestorRecursive(root->right, p, q);
        if (left == NULL) {
            return right;
        } else if (right == NULL) {
            return left;
        } else {
            return root;
        }
    }

    // Approach 2: Iterative using Parent Pointers
    // - Use a stack to traverse the tree and store each node's parent.
    // - Store the paths from the root to p and q in separate vectors.
    // - Compare the paths to find the last common node.
    TreeNode* lowestCommonAncestorIterativeParentPointers(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (!root) return nullptr;

        std::map<TreeNode*, TreeNode*> parentMap; // Map to store child -> parent mapping
        std::stack<TreeNode*> stack;
        stack.push(root);
        parentMap[root] = nullptr; // Root has no parent

        // Iterate until we find both p and q
        while (parentMap.find(p) == parentMap.end() || parentMap.find(q) == parentMap.end()) {
            TreeNode* node = stack.top();
            stack.pop();

            if (node->left) {
                parentMap[node->left] = node;
                stack.push(node->left);
            }
            if (node->right) {
                parentMap[node->right] = node;
                stack.push(node->right);
            }
        }

        // Get the ancestors of p
        std::vector<TreeNode*> pathP;
        while (p) {
            pathP.push_back(p);
            p = parentMap[p];
        }

        // Get the ancestors of q
        std::vector<TreeNode*> pathQ;
        while (q) {
            pathQ.push_back(q);
            q = parentMap[q];
        }

        // Find the last common ancestor
        TreeNode* lca = nullptr;
        for (int i = pathP.size() - 1, j = pathQ.size() - 1; i >= 0 && j >= 0; --i, --j) {
            if (pathP[i] == pathQ[j]) {
                lca = pathP[i];
            } else {
                break;
            }
        }
        return lca;
    }

    // Approach 3: Using Node Levels (Iterative)
    // - Find the levels of nodes p and q.
    // - Bring the deeper node to the same level as the shallower node.
    // - Move both nodes up until they meet.
    TreeNode* lowestCommonAncestorWithLevels(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (!root) return nullptr;

        std::map<TreeNode*, int> levelMap; // Stores node -> level
        std::queue<std::pair<TreeNode*, int>> queue; // Stores node and its level
        queue.push({root, 0});
        levelMap[root] = 0;

        // BFS to calculate levels of all nodes, including p and q
        while (!queue.empty()) {
            TreeNode* node = queue.front().first;
            int level = queue.front().second;
            queue.pop();

            if (node->left) {
                levelMap[node->left] = level + 1;
                queue.push({node->left, level + 1});
            }
            if (node->right) {
                levelMap[node->right] = level + 1;
                queue.push({node->right, level + 1});
            }
        }
        int levelP = levelMap[p];
        int levelQ = levelMap[q];

        // Bring the deeper node to the same level as the shallower node
        while (levelP > levelQ) {
            p = findParent(root, p); // Helper to find parent, see below
            levelP--;
        }
        while (levelQ > levelP) {
            q = findParent(root, q); // Helper to find parent
            levelQ--;
        }

        // Move both nodes up until they meet
        while (p != q) {
            p = findParent(root, p); // Helper to find parent
            q = findParent(root, q); // Helper to find parent
        }
        return p;
    }

    // Helper function to find the parent of a node in the tree.  Used by approach 3.
    TreeNode* findParent(TreeNode* root, TreeNode* child) {
        if (!root || root == child) return nullptr;
        std::queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()) {
            TreeNode* node = q.front();
            q.pop();
            if (node->left == child || node->right == child) {
                return node;
            }
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }
        return nullptr;
    }



    // Approach 4: Morris Traversal (No Extra Space)
    // - Morris traversal is a way to traverse a tree without using a stack or recursion.
    // - It works by temporarily modifying the tree structure.
    // - Find p and q using Morris Traversal.
    // - After finding p and q, find LCA.
    TreeNode* lowestCommonAncestorMorris(TreeNode* root, TreeNode* p, TreeNode* q) {
        TreeNode* first = nullptr;
        TreeNode* second = nullptr;
        TreeNode* current = root;
        TreeNode* pre;

        // Find p and q using Morris Traversal
        while (current && (first == nullptr || second == nullptr)) {
            if (current->left == nullptr) {
                if (current == p) first = p;
                if (current == q) second = q;
                current = current->right;
            } else {
                pre = current->left;
                while (pre->right != nullptr && pre->right != current)
                    pre = pre->right;

                if (pre->right == nullptr) {
                    pre->right = current;
                    if (current == p) first = p;
                    if (current == q) second = q;
                    current = current->left;
                } else {
                    pre->right = nullptr;
                    current = current->right;
                }
            }
        }

        if (first == nullptr || second == nullptr) return nullptr; // Handle if p or q not in tree

        // Find LCA
        std::vector<TreeNode*> pathP, pathQ;
        getPath(root, p, pathP);
        getPath(root, q, pathQ);

        TreeNode* lca = nullptr;
        for (size_t i = 0, j = 0; i < pathP.size() && j < pathQ.size(); ++i, ++j) {
            if (pathP[i] == pathQ[j])
                lca = pathP[i];
            else
                break;
        }
        return lca;
    }

    // Helper function to get path from root to a node. Used by Morris.
    bool getPath(TreeNode* root, TreeNode* target, std::vector<TreeNode*>& path) {
        if (!root) return false;
        path.push_back(root);
        if (root == target) return true;
        if (getPath(root->left, target, path) || getPath(root->right, target, path))
            return true;
        path.pop_back();
        return false;
    }


    // Approach 5: Iterative using Stack
    // - Use a stack to keep track of the path from the root to the current node.
    // - When we find p or q, save the current path.
    // - Compare the two paths to find the LCA.
    TreeNode* lowestCommonAncestorIterativeStack(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (!root) return nullptr;

        std::stack<TreeNode*> stack;
        std::vector<TreeNode*> pathP, pathQ;
        std::map<TreeNode*, bool> visited; // Keep track of visited

        stack.push(root);
        visited[root] = true;

        while (!stack.empty() && (pathP.empty() || pathQ.empty())) {
            TreeNode* current = stack.top();

            if (current == p && pathP.empty()) {
                // Convert stack to vector
                std::stack<TreeNode*> tempStack = stack;
                while (!tempStack.empty()) {
                    pathP.push_back(tempStack.top());
                    tempStack.pop();
                }
                std::reverse(pathP.begin(), pathP.end()); // Reverse to get correct order
            }
            if (current == q && pathQ.empty()) {
                 // Convert stack to vector
                std::stack<TreeNode*> tempStack = stack;
                while (!tempStack.empty()) {
                    pathQ.push_back(tempStack.top());
                    tempStack.pop();
                }
                std::reverse(pathQ.begin(), pathQ.end());  // Reverse to get correct order
            }

            // Explore left or right if possible
            if (current->left && !visited[current->left]) {
                stack.push(current->left);
                visited[current->left] = true;
            } else if (current->right && !visited[current->right]) {
                stack.push(current->right);
                visited[current->right] = true;
            } else {
                stack.pop(); // Backtrack
            }
        }
        // Find LCA
        TreeNode* lca = nullptr;
        for (size_t i = 0; i < pathP.size() && i < pathQ.size(); ++i) {
            if (pathP[i] == pathQ[i]) {
                lca = pathP[i];
            } else {
                break;
            }
        }
        return lca;
    }
};

// Helper function to create a sample binary tree for testing
TreeNode* createSampleTree() {
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(5);
    root->right = new TreeNode(1);
    root->left->left = new TreeNode(6);
    root->left->right = new TreeNode(2);
    root->right->left = new TreeNode(0);
    root->right->right = new TreeNode(8);
    root->left->right->left = new TreeNode(7);
    root->left->right->right = new TreeNode(4);
    return root;
}

int main() {
    Solution solution;
    TreeNode* root = createSampleTree();
    TreeNode* p = root->left; // Node 5
    TreeNode* q = root->right->right; // Node 8

    // Test all 5 approaches
    std::cout << "Lowest Common Ancestor (LCA) of " << p->val << " and " << q->val << " is:" << std::endl;

    // Approach 1: Recursive Approach
    TreeNode* lca1 = solution.lowestCommonAncestorRecursive(root, p, q);
    std::cout << "Recursive Approach: " << (lca1 ? lca1->val : -1) << std::endl;

    // Approach 2: Iterative using Parent Pointers
    TreeNode* lca2 = solution.lowestCommonAncestorIterativeParentPointers(root, p, q);
    std::cout << "Iterative with Parent Pointers: " << (lca2 ? lca2->val : -1) << std::endl;

    // Approach 3: Using Node Levels (Iterative)
    TreeNode* lca3 = solution.lowestCommonAncestorWithLevels(root, p, q);
    std::cout << "Iterative with Levels: " << (lca3 ? lca3->val : -1) << std::endl;

    // Approach 4: Morris Traversal (No Extra Space)
    TreeNode* lca4 = solution.lowestCommonAncestorMorris(root, p, q);
    std::cout << "Morris Traversal: " << (lca4 ? lca4->val : -1) << std::endl;

    // Approach 5: Iterative using Stack
    TreeNode* lca5 = solution.lowestCommonAncestorIterativeStack(root, p, q);
    std::cout << "Iterative with Stack: " << (lca5 ? lca5->val : -1) << std::endl;

    return 0;
}
